File: Python/data_structures/02_tree_height.py
import sys


def convert_to_intlist(line):
    return [int(item) for item in line.split()]


class Node(object):
    def __init__(self, elem):
        self._key = elem
        self._children = set()

    @property
    def key(self):
        return self._key

    @property
    def children(self):
        return self._children

    def add_child(self, key):
        self._children.add(key)


class Tree(object):
    def __init__(self, number_of_nodes, parents_of_nodes):
        self._root = None
        self._nodes = self._build_tree(parents_of_nodes, number_of_nodes)

    def _build_tree(self, parents_of_nodes, number_of_nodes):
        nodes = [Node(k) for k in range(number_of_nodes)]
        for node_id, parent_id in enumerate(parents_of_nodes):
            if parent_id == -1:
                self._root = nodes[node_id]
            else:
                nodes[parent_id].add_child(node_id)
        return nodes

    def _get_children_of_nodes(self, keys):
        children = set()
        for key in keys:
            children.update(self._nodes[key].children)
        return children

    def calc_depth(self):
        new_keys = {self._root.key}
        depth = 0
        while new_keys:
            depth += 1
            new_keys = self._get_children_of_nodes(new_keys)
        return depth


def main():
    reader = sys.stdin
    number_of_nodes = int(next(reader))
    parent_ids = convert_to_intlist(next(reader))
    tree = Tree(number_of_nodes, parent_ids)
    depth = tree.calc_depth()
    print(depth)


if __name__ == "__main__":
    main()

File: Haskell/data_structures/02-TreeHeight.hs
import Data.Map.Strict (Map(..))
import Data.Vector (Vector(..), (!), generate)
import Data.Set (Set(..))
import qualified Data.Map.Strict as M
import qualified Data.Set as S

data Node = Node { key :: Int, children :: Set Int }
data Tree = Tree { root :: Node, nodes :: Vector Node }


convertToIntList :: String -> [Int]
convertToIntList = map read . words


getIndexOfRoot :: Map Int (Set Int) -> Int
getIndexOfRoot childrenOfNodes = head $ S.elems $ childrenOfNodes M.! (-1)


calcChildrenFromParents :: Int -> [Int] -> Map Int (Set Int)
calcChildrenFromParents nrNodes parentsOfNodes = foldl update M.empty parentsWithIndices
    where
        parentsWithIndices = zip parentsOfNodes [0..]
        update :: Map Int (Set Int) -> (Int, Int) -> Map Int (Set Int)
        update table (parentId, nodeId) =
            let children = M.findWithDefault S.empty parentId table
                children' = S.insert nodeId children
            in M.insert parentId children' table


buildTree :: Int -> [Int] -> Tree
buildTree nrNodes parentsOfNodes =
    let childrenOfNodes = calcChildrenFromParents nrNodes parentsOfNodes
        nodes = generate nrNodes (\key -> Node key (M.findWithDefault S.empty key childrenOfNodes))
        indexOfRoot = getIndexOfRoot childrenOfNodes
    in Tree (nodes ! indexOfRoot) nodes


getChildrenOfNodes :: Set Int -> Vector Node -> Set Int
getChildrenOfNodes keys nodes = S.foldl (\acc key -> S.union acc (extractChildren key)) S.empty keys
    where
        extractChildren :: Int -> Set Int
        extractChildren key = children $ nodes ! key


calcTreeDepth :: Tree -> Int
calcTreeDepth (Tree root nodes) = go 0 (S.singleton $ key root)
    where
        go :: Int -> Set Int -> Int
        go depth keys
            | S.null keys = depth
            | otherwise = let nextKeys = getChildrenOfNodes keys nodes in go (depth + 1) nextKeys


main :: IO()
main = do
    nrNodes <- readLn
    parentsOfNodes <- fmap convertToIntList getLine
    let tree = buildTree nrNodes parentsOfNodes
    let depth = calcTreeDepth tree
    print depth
